lib/stdlib/src/erl_lint.erl

```diff
@@ -3020,80 +3032,108 @@ check_type(Types, St) ->
 		      AccSt
 	      end, St1, SeenVars).
 
-check_type({ann_type, _A, [_Var, Type]}, SeenVars, St) ->
-    check_type(Type, SeenVars, St);
-check_type({remote_type, A, [{atom, _, Mod}, {atom, _, Name}, Args]},
+check_type_1({type, Anno, TypeName, any}=Type, SeenVars, #lint{types=Types}=St) ->
+    TypePair = {TypeName, 0},
+    case is_map_key(TypePair, Types) of
+        true ->
+            check_type_2(Type, SeenVars, used_type(TypePair, Anno, St));
+        false ->
+            check_type_2(Type, SeenVars, St)
+    end;
+check_type_1({type, Anno, TypeName, Args}=Type, SeenVars, #lint{types=Types}=St) ->
+    TypePair = {TypeName, length(Args)},
+    case is_map_key(TypePair, Types) of
+        true ->
+            check_type_2(Type, SeenVars, used_type(TypePair, Anno, St));
+        false ->
+            check_type_2(Type, SeenVars, St)
+    end;
+check_type_1({TypeName, Anno, Args}=Type, SeenVars, #lint{types=Types}=St) when is_list(Args) ->
+    TypePair = {TypeName, length(Args)},
+    case is_map_key(TypeName, Types) of
+        true ->
+            TypePair = {TypeName, length(Args)},
```

Comment:
The third clause (the one with the redundant definition) is wrong and redundant. It handles annotated types (e.g. `A :: atom()`), remote types (e.g. `gb_sets:set()`), and integer and atom literals. Those need no special handling, so the third clause can be removed.

The first and second clauses can indeed be combined to avoid repeating code.

I have pushed a fixup commit.